include/dlaf/eigensolver/band_to_tridiag/mc.h

```diff
@@ -365,6 +509,126 @@ class SweepWorker {
+            MPI_Request* req) const noexcept {
+    DLAF_MPI_CHECK_ERROR(MPI_Isend(const_cast<T*>(data_()), to_int(band_size_ + 1),
+                                   comm::mpi_datatype<T>::type, dest, tag, comm, req));
+  }
+
+  void recv(SizeType sweep, SizeType step, const comm::Communicator& comm, comm::IndexT_MPI src,
+            comm::IndexT_MPI tag, MPI_Request* req) noexcept {
+    SweepWorker<T>::setId(sweep, step);
+    *data_(0) = T{9};
+    *data_(1) = T{9};
+    DLAF_MPI_CHECK_ERROR(
+        MPI_Irecv(data_(), to_int(band_size_ + 1), comm::mpi_datatype<T>::type, src, tag, comm, req));
+  }
+
+  using SweepWorker<T>::compactCopyToTile;
+
+private:
+  void doStepSplit(BandBlockDist<T>& a, SizeType n1) noexcept {
+    SizeType j = this->firstRowHHR();
+    SizeType n = this->sizeHHR();  // size diagonal tile and width off-diag tile
+    SizeType m = std::min(band_size_, size_ - band_size_ - j);  // height off diagonal tile
+    const auto n2 = n - n1;
+
+    applyHHLeftRightHerm(n1, n2, tau(), v(), a.ptr(0, j), a.ptr(0, j + n1), a.ld(), w());
+    if (m > 0) {
+      applyHHRight(m, n1, n2, tau(), v(), a.ptr(n, j), a.ptr(n2, j + n1), a.ld(), w());
+    }
+    if (m > 1) {
+      HHReflector(m, tau(), v(), a.ptr(n, j));
+      applyHHLeft(m, n1 - 1, n2, tau(), v(), a.ptr(n - 1, j + 1), a.ptr(n2, j + n1), a.ld(), w());
+    }
+    step_ += 1;
+    // Note: the sweep is completed if m <= 1.
+  }
+
+  using SweepWorker<T>::size_;
+  using SweepWorker<T>::band_size_;
+  using SweepWorker<T>::sweep_;
+  using SweepWorker<T>::step_;
+  using SweepWorker<T>::data_;
+  using SweepWorker<T>::tau;
+  using SweepWorker<T>::v;
+  using SweepWorker<T>::w;
+};
+
+template <class CommSender, class PromiseSender>
+[[nodiscard]] auto scheduleSendWorker(CommSender&& pcomm, comm::IndexT_MPI dest, comm::IndexT_MPI tag,
+                                      PromiseSender&& worker) {
+  using dlaf::comm::internal::transformMPI;
+  using dlaf::internal::whenAllLift;
+
+  auto send = [dest, tag](const comm::Communicator& comm, const auto& worker, MPI_Request* req) {
+    worker.send(comm, dest, tag, req);
+  };
+
+  return whenAllLift(std::forward<CommSender>(pcomm), std::forward<PromiseSender>(worker)) |
+         transformMPI(send);
+}
+
+template <class CommSender, class PromiseSender>
+[[nodiscard]] auto scheduleRecvWorker(SizeType sweep, SizeType step, CommSender&& pcomm,
+                                      comm::IndexT_MPI src, comm::IndexT_MPI tag,
+                                      PromiseSender&& worker) {
+  using dlaf::comm::internal::transformMPI;
+  using dlaf::internal::whenAllLift;
+
+  auto recv = [sweep, step, src, tag](const comm::Communicator& comm, auto& worker, MPI_Request* req) {
+    worker.recv(sweep, step, comm, src, tag, req);
+  };
+
+  return whenAllLift(std::forward<CommSender>(pcomm), std::forward<PromiseSender>(worker)) |
+         transformMPI(recv);
+}
+
+template <class T>
+class vector2D {
```

Comment:
Am I missing something or is this unused? I guess you might (have) intend(ed) to use it to replace the `vector<vector<T>>` uses further down?

Reply:
It was a first attempt, but then switched to `vector<vector<...>>` as the "internal" vector need to be resized.